docs/routing.md

```diff
@@ -0,0 +1,248 @@
+
+The `resource` phase handles any remaining user-specified rewrites, and then checks the build output for any remaining routes. Further, it sets the status to 404 for any other routes that have not had a match yet.
+
+- `next.config.js`
+  - Applying `fallback` [`rewrites`](https://nextjs.org/docs/pages/api-reference/next-config-js/rewrites).
+
+### `miss`
+
+At this point, we have checked all possible routes, and none of them have matched. This ultimately means that the path could not be found in the build output at any point, and so we need to return a 404 response to the client.
+
+### `error`
+
+The error phase is a rather interesting one in that it appears to need completely custom behaviour to function properly. It defines a list of source routes with status codes, usually to map a request to a custom error page.
+
+The unusual part, however, is that there is no clear way to differentiate between which should be matched, without changing the default behaviour. Instead of treating the status code property as something to be applied to the response, instead we have to treat it as something that we are trying to match with.
+
+### `hit`
+
+At the end of matching, we finally enter the `hit` phase. Here, additional headers are applied, such as the `x-matched-path` header. We enter this phase regardless of if the match ultimately resulted in a miss or not.
+
+Normally, when a match results in a miss, it will use a page at `/404` to handle the response. This would still need to have the relevant headers applied from _hitting_ a file.
+
+## Routing Process
+
+Now that we have a better understanding of the different phases, we can look at how the routing process works as a whole. The actual system that was designed to handle this is a little complex, but it can be broken down into a series of steps.
+
+Along the way, there are several times where custom behaviour has had to be introduced. This is, in part, to maximise the amount of features we support, but also because the routing system used internally by Vercel is not necessarily documented that much. Our system is custom built to serve requests, designed around our understanding of how Vercel's system should function with the Vercel build output.
+
+A simplistic overview of the routing process is as follows.
+
+- Call the handler with the request.
+- Check each phase from the build output configuration.
+  - Check each source route in a phase to see if it matches.
+  - If a match is found, apply any relevant response modifiers.
+  - Determine what to do next; return the match, or continue through routing.
+- Run the matched function, or fetch the matched static asset.
+- Return the final response.
+
+### Checking Phases
+
+The first thing that happens when a request is received, is that the router starts looking for matches in the `none` phase. It calls a function to start checking phases. This function is called recursively to check additional phases.
```

Comment:
sorry I find this sentence kind of unclear, I think I do get the point but I feel like it could be presented better in a simpler way

mh.... now that I think about it, is this really necessary? isn't this describing the implementation too much into details? (like I care that we check each phase but I don't necessarily want to know if there is a function that does that and if the function is called recursively, especially since that is one of those things likely things to change with refactoring and whatnot)